### PR TITLE
Feature/justinyaodu/1250719428 edit user route

### PR DESCRIPTION
**Pull request by justinyaodu**
### Administrative Info
Monday Board ID: 1181559666
Make sure your branch name conforms to: `<feature/staging/hotfix/...>/[username]/[Monday Item ID]-[3-4 word description separated by dashes]`. Otherwise, please rename your branch and create a new PR.

### Changes
What changes did you make?
- Defined what fields can be edited by admins and non-admins
- Added route to edit users
- Prevent refresh token from being included when User objects are converted to JSON

### Testing
How did you confirm your changes worked? 

Ran the following script, which tests every combination of {developer, president} editing {developer, president}'s {email, Discord username}:

```bash
#!/bin/bash

backend='localhost:8000'

content_type_json='Content-Type: application/json; charset=utf-8'

dev_creds='{"email": "developer@example.com", "password": "password"}'
pres_creds='{"email": "president@example.com", "password": "password"}'

function get_login() {
	curl -H "${content_type_json}" -d "${1}" "${backend}/api/auth/login"
}

function get_token() {
	jq -r '.token' <<< "${1}"
}

function get_id() {
	jq -r '.user._id' <<< "${1}"
}

dev_login="$(get_login "${dev_creds}")"
pres_login="$(get_login "${pres_creds}")"

dev_token="$(get_token "${dev_login}")"
pres_token="$(get_token "${pres_login}")"

dev_id="$(get_id "${dev_login}")"
pres_id="$(get_id "${pres_login}")"

function curl_json() {
	local cmd=(curl -H "${content_type_json}" "${@}")
	echo 'Request:'
	echo "${cmd[@]@Q}"

	echo 'Response:'
	"${cmd[@]}"
	echo

	echo
}

function change_discord() {
	curl_json -X PUT -H "Authorization: Bearer ${1}" -d '{"_id": "'"${2}"'", "discord_username": "'"foo${RANDOM}#1234"'"}' "${backend}/api/users"
}

function change_email() {
	curl_json -X PUT -H "Authorization: Bearer ${1}" -d '{"_id": "'"${2}"'", "email": "'"email${RANDOM}@example.com"'"}' "${backend}/api/users"
}

echo "Should succeed: dev changes own Discord username"
change_discord "${dev_token}" "${dev_id}"

echo "Should fail: dev changes president's Discord username"
change_discord "${dev_token}" "${pres_id}"

echo "Should succeed: president changes dev's Discord username"
change_discord "${pres_token}" "${dev_id}"

echo "Should succeed: president changes own Discord username"
change_discord "${pres_token}" "${pres_id}"

echo "Should fail: dev changes own email"
change_email "${dev_token}" "${dev_id}"

echo "Should fail: dev changes president's email"
change_email "${dev_token}" "${pres_id}"

echo "Should succeed: president changes dev's email"
change_email "${pres_token}" "${dev_id}"

echo "Should succeed: president changes own email"
change_email "${pres_token}" "${pres_id}"
```

### Confirmation of Change 
Upload a screenshot, if possible. Otherwise, please provide instructions on how to see the change.

| Running the test script after clearing the database |
| --- |
| image |

### CODE REVIEW

File: backend/models/user.js [hunk 1]
```diff
@@ -82,6 +82,7 @@ UserSchema.set("toJSON", {
   transform(doc, obj) {
     const ret = { ...obj };
     delete ret.password;
+    delete ret.refreshToken;
```
Comment (AnsonLee99) [inline backend/models/user.js:85]: What is the purpose of deleting the refresh token here?

**Review by AnsonLee99 — APPROVED** [review]
Just a quick question on user.js that I commented. Otherwise, lgtm!

**Review by lynguyxn — APPROVED** [review]
lgtm

File: backend/models/user.js (hunk 1 above)
Comment (justinyaodu) [reply backend/models/user.js:85]: I realized that the route which gets all users actually returns the refresh token for every user, which is rather problematic from a security perspective. This change makes sure that the refresh token doesn't show up when a User object is converted to JSON. This change is mostly unrelated to the rest of the PR.

File: backend/routes/users.js [hunk 2]
```diff
@@ -18,4 +18,17 @@ router.get("/", [authorizeUser(["roster"])], (req, res, next) => {
     });
 });
 
+/**
+ * Edit user information
+ */
+router.post("/", [authorizeUser([])], (req, res, next) => {
```
Comment (hahsu) [inline backend/routes/users.js:24]: Should this be a put request since we are editing user information?

**Review by hahsu — CHANGES_REQUESTED** [review]

File: backend/routes/users.js (hunk 2 above)
Comment (justinyaodu) [reply backend/routes/users.js:24]: I agree! Thanks for catching that. I'll make that change now.

**Review by hahsu — APPROVED** [review]
Lgtm!

**Pull request merged**